man/man1/newspost.1
.TH "NEWSPOST" "1" "2.1" "Jim Faulkner" ""
.SH "NAME"
.LP 
newspost \- a usenet binary autoposter
.SH "SYNTAX"
.LP 
newspost [\fIoptions\fP [\fIarguments\fP]] <\fIfilenames\fP>
.SH "DESCRIPTION"
.LP 
Newspost is a usenet binary autoposter.  It will uuencode or yencode files
and post them to usenet newsgroups. In addition, it can automatically
generate and post .SFV checksum files and .PAR parity files.
.SH "OPTIONS"
.LP 
.TP 
\fB\-i\fR <\fIaddress\fP>
The news server to post to.  <\fIaddress\fP> must be either a hostname or
an IP address.  If the NNTPSERVER environment variable is set, this is set
to that value by default.
.TP 
\fB\-z\fR <\fInumber\fP>
Sets the port on which to connect to <\fInumber\fP>.
.TP 
\fB\-u\fR <\fIstring\fP>
Sets your username on the news server to <\fIstring\fP>.
.TP 
\fB\-p\fR <\fIstring\fP>
Sets your password on the news server to <\fIstring\fP>.
.TP 
\fB\-f\fR <\fIaddress\fP>
Your e\-mail address.  <\fIaddress\fP> must be a real e\-mail address, or
your posts may fail.  If the USER and HOSTNAME environment variables are
set, this is set to $USER@$HOSTNAME by default.
.TP 
\fB\-F\fR <\fIstring\fP>
Sets your name in the From header line to <\fIstring\fP>.  Please note:  
if your name has spaces in it, it should be enclosed in quotes.
.TP 
\fB\-o\fR <\fIstring\fP>
Sets the Organization header line to <\fIstring\fP>.  Please note:  if
your organization has spaces in it, it should be enclosed in quotes.
.TP 
\fB\-n\fR <\fIstring\fP>
Sets the newsgroup to post to.  <\fIstring\fP> may also be a comma
separated list of newsgroups, but you may only post to at most 5
newsgroups.
.TP 
\fB\-s\fR <\fIstring\fP>
Sets <\fIstring\fP> as the Subject header line.  This subject line is
always appended with ' \- filename (1/9)' (depending on the filename, part
number, and number of parts).  Other options may add or change information
in the subject line (see those options for details).  Please note:  if
your subject has spaces in it, it should be enclosed in quotes.  See
EXAMPLES for more information.
.TP 
\fB\-w\fR <\fIstring\fP>
Sets the Followup\-To header line.  <\fIstring\fP> must be a newsgroup.
.TP 
\fB\-m\fR <\fIaddress\fP>
Sets the Reply\-To header line.  <\fIaddress\fP> must be an e\-mail address.
.TP 
\fB\-r\fR <\fIstring\fP>
Set the References header line.  <\fIstring\fP> must be a message\-ID or a
comma\-separated list of message\-IDs.
.TP 
\fB\-x\fR
By default all posts include the "X\-No\-Archive: yes" header.  This
option DISABLES that behaviour.
.TP 
\fB\-X\fR <\fIstring\fP>
Sets <\fIstring\fP> as a complete header line.  This option may be
specified multiple times to add more header lines.  Please note:  do not
use this feature unless you know exactly what you are doing.
.TP 
\fB\-q\fR
This specifies that ' \- File 01 of 10: ' (depending on the file number and
number of files) will be added to the subject line after text specified by
the \-s switch, and before the file information.  If the file being posted
is a generated .SFV file, then the string is ' \- SFV File 1 of 1: '.  If
the file is a generated .PAR file, the string is ' \- PAR File 01 of 10: '
(depending on the .PAR file number and the number of .PAR files).  See
EXAMPLES for more information.
.TP 
\fB\-y\fR
Use yencoding instead of uuencoding.  Yencoding results in smaller, faster
posts, but most news readers do not support yencoding.  When yencoding is
enabled, ' \- "filename" yEnc (1/9)' is appended to the subject line
instead of what is specified above for the \-s option.  Please see
http://www.yenc.org for more information about yencoding.
.TP 
\fB\-e\fR <\fIfilename\fP>
Posts <\fIfilename\fP> as a text file (no encoding) before the first file
posted.  This file is posted with the exact subject line as the first file
posted, except that (0/9) is displayed as the part number (depending on
the number of parts in the first file posted).  As an alternative, "\-0" 
(zero, not 'O') may be used instead of "\-e".
.TP 
\fB\-E\fR
Opens the editor set by the EDITOR environment variable to create a
temporary text file which is posted as a prefix, like a file specified
with the \-e option.  If EDITOR is not set then vi is used.
.TP 
\fB\-c\fR <\fIfilename\fP>
Generates and posts a .SFV checksum file named <\fIfilename\fP>.  If
<\fIfilename\fP> does not end in '.sfv', it will automatically be appended
to the filename.  Note the change from newspost 1.x, which used "\-v" for 
this option; also, the .SFV file is now posted first instead of last.
.TP 
\fB\-a\fR <\fIfilename\fP>
Generates and posts .PAR files whose name is based on <\fIfilename\fP>.  
If <\fIfilename\fP> does not end in '.par', it will automatically be 
appended to the filename.  The .PAR header will end in '.par', and 
each .PAR volume will end in '.p01', '.p02', etcetera.  Please see
http://parchive.sourceforge.net for more information about .PAR files.
.TP 
\fB\-A\fR <\fInumber\fP>
Sets the number of .PAR volumes to create to <\fInumber\fP>.  This is in
addition to the .PAR header file.  The \-a option must be used for this
option to do anything.
.TP 
\fB\-B\fR <\fInumber\fP>
Sets the number of .PAR volumes to create based on the number of files.  
For every <\fInumber\fP> files, newspost will create one .PAR volume.  
This is in addition to the .PAR header file.  By default, newspost creates
one .PAR volume for every ten files posted.
.TP 
\fB\-l\fR <\fInumber\fP>
Sets the number of lines per message to <\fInumber\fP>.  Most people post
messages which are between 5000 and 10000 lines long.  By default, this is
set to 7500.  Note: For uuencoded messages, this is the actual number of 
lines in the body of the message; but for yencoded messages, it's used to 
determine the size of each segment before encoding, by multiplying the 
specified number of lines by 45 (which is the size of a uuencoded line 
before encoding). Thus, the size of each segment before encoding is the
same for either method, but the actual line count for yencoded segments 
will vary.
.TP 
\fB\-t\fR
When this option is specified, one file may be posted as a plain text
message.  If no file is specified, EDITOR (or vi) is opened to create the
message.
.TP 
\fB\-T\fR <\fInumber\fP>
Sets the time to wait before posting to <\fInumber\fP>.  This number is
interpreted in seconds by default, but you can specify, for example, "3m"
if you want 3 minutes, or "1h" if you want 1 hour.  This number can also
be specified in hours:minutes:seconds or minutes:seconds format.  e.g.
"5:30" if you want to wait 5 minutes 30 seconds, or "4:0:0" if you want to
wait 4 hours.  This value is set to 10 seconds by default, and may not be
set to less than 3 seconds.
.TP 
\fB\-k\fR <\fIdirname\fP>
Sets the directory to be used for storing temporary files to
<\fIdirname\fP>.  <\fIdirname\fP> must exist. If the TMPDIR environment
variable is set, this is set to that value by default.  If the TMP
environment variable is set (and TMPDIR is not), this is set to that value
by default.
.TP 
\fB\-d\fR
Sets the current options as default by saving them to $HOME/.newspostrc.  
Only the following options may be saved as default: \-i, \-z, \-u, \-p,
\-f, \-F, \-o, \-n, \-w, \-m, \-x, \-q, \-y, \-B, \-l, \-X, and \-k.
.TP 
\fB\-D\fR <\fIoption\fP>
This option disables or clears other options.  When y, q, or x is passed
as <\fIoption\fP>, it has the OPPOSITE effect of either the \-y, \-q, or
\-x option, respectively.  It will clear the corresponding options for u,
p, F, o, w, and m.  It also takes X as an argument, but it clears ALL
extra headers set with the \-X option.  This option may be specified
multiple times with different values for <\fIoption\fP>.  Please note only
options that are not required can be cleared.
.TP 
\fB\-v\fR
When this option is specified, newspost will be extra verbose.  (Note that 
"\-v" was used for .SFV generation in newspost 1.x.)
.TP 
\fB\-V\fR
Print version information and exit.
.TP 
\fB\-h\fR
Print help and exit.  This is the default when there are no arguments.
.SH "PARTIAL AND FAKE POSTING"
Parts of files may be posted by specifying filenames in the following
format: 'filename:partnumbers'.  Colons in the first or second position 
(as after a drive letter) are ignored.  Partnumbers is a comma\-separated list of
parts to post.  Each item in the list may be any of three formats: '2' (a
single part), '2\-5' (parts 2, 3, 4, and 5), or '2+' (part 2 and all parts
up to and including the last part).  If, for example, you want to post
parts 2, 4, 5, 6, 10, 11, and 12 of track01.mp3 (a 12 part file), you
could specify the filename like this: 'track01.mp3:2,4\-6,10+'.  If there
is no ':' in a specified filename, newspost will post the entire file. If
the number 0 is specified after the colon, newspost will add the file to generated .SFV
and .PAR files, and preserve ' \- File x of y: ' numbering, but not
actually post the file.  See EXAMPLES for more information.  Please note:  If your files
did not post completely, your file is much more likely to post properly if you repost 
the entire file, or split the file into smaller parts and post them.
.SH "SPEED INDICATOR"
While posting, Newspost shows the approximate upload speed in bytes per 
second.  This is measured by the total number of \fIencoded\fP bytes sent 
to the server.  In Newspost 1.x, the speed indicator was based on the number 
of \fIunencoded\fP bytes, so this speed indicator will show a higher number, 
though the speed is the same.  If you are connecting to your news server via 
a compressed link, you may see a difference in transfer rates when using 
uuencoding vs. yencoding.
.SH "FILES"
.LP 
\fI$HOME/.newspostrc\fP is an optional file used to store defaults. 
Newspost will also read (but not write) the old\-style .newspost file if it 
is present and .newspostrc is not.
.SH "ENVIRONMENT VARIABLES"
.LP 
.TP 
\fBHOME\fP
Specifies your home directory.  Newspost will not work properly if this is
not set.
.TP 
\fBNNTPSERVER\fP
Specifies the hostname or IP address of your news server.
.TP 
\fBUSER\fP
Used in creating the default from address if HOSTNAME is set.
.TP 
\fBHOSTNAME\fP
Used in creating the default from address if USER is set.
.TP 
\fBTMPDIR\fP
Specifies the directory used for creating temporary files.
.TP 
\fBTMP\fP
Specifies the directory used for creating temporary files if TMPDIR is not
set.
.TP 
\fBEDITOR\fP
Specifies the text editor to use.
.SH "EXAMPLES"
.LP 
.na
.TP 
\fBSave your news server, e\-mail address, and name as default:\fP
.br 
newspost \-d \-i news.myisp.com \-f newspost@unixcab.org \-F 'Jim Faulkner'
.TP 
\fBPost some files to alt.binaries.test:\fP
.br 
newspost \-n alt.binaries.test \-s 'Here are some songs' /nfs/music/*.mp3
.TP 
\fBA subject line from the above post may look like this:\fP
.br 
Here are some songs \- track01.mp3 (01/15)
.TP 
\fBPost some files to alt.binaries.test.yenc using yencoding:\fP
.br 
newspost \-y \-n alt.binaries.test.yenc \-s 'Here are some more songs'
/nfs/music/*.mp3
.TP 
\fBA subject line from the above post may look like this:\fP
.br 
Here are some more songs \- "track01.mp3" yEnc (01/12)
.TP 
\fBInclude "File x of y" in the subject:\fP
.br 
newspost \-q \-y \-n alt.binaries.test.yenc \-s 'Here are some more songs'
track01.mp3 track02.mp3 track03.mp3
.TP 
\fBA subject line from the above post may look like this:\fP
.br 
Here are some more songs \- File 1 of 3: "track01.mp3" yEnc (01/12)
.TP 
\fBMake yencode and "File x of y" the default:\fP
.br 
newspost \-d \-q \-y
.TP 
\fBAdd extra headers to every post by default:\fP
.br 
newspost \-d \-X 'X\-My\-Stupid\-Header\-1' \-X 'X\-My\-Stupid\-Header\-2'
.TP 
\fBClear extra headers and "File x of y", set uuencode, default:\fP
.br 
newspost \-d \-DX \-Dq \-Dy
.TP 
\fBGenerate and include an SFV file:\fP
.br 
newspost \-c checksum.sfv \-n alt.binaries.test \-s 'Here are some songs'
/nfs/music/*.mp3
.TP 
\fBGenerate and include PAR files:\fP
.br 
newspost \-a parfile.par \-n alt.binaries.test \-s 'Here are some songs'
/nfs/music/*.mp3
.TP 
\fBCreate a PAR volume for every 7 files posted by default:\fP
.br 
newspost \-d \-B 7
.TP 
\fBPost with 3 PAR volumes:\fP
.br 
newspost \-q \-a parfile.par \-A 3 \-n alt.binaries.test \-s 'Here are
some songs' /nfs/music/*.mp3
.TP 
\fBA subject line from the above post may look like this:\fP
.br 
Here are some songs \- PAR File 1 of 4: parfile.par (1/1)
.TP 
\fBPost parts 3, 4, and 5 of track01.mp3 to alt.binaries.test:\fP
.br 
newspost \-n alt.binaries.test \-s 'Here are some songs' track01.mp3:3\-5
.TP 
\fBPost all of track01.mp3, parts of track02.mp3 and track03.mp3:\fP
.br 
newspost \-n alt.binaries.test \-s 'Here are some songs' track01.mp3
track02.mp3:11+ track03.mp3:1,3,6\-8
.TP 
\fBResume a broken post at Track02.mp3 preserving -q numbering:\fP
.br 
newspost \-q \-s '2 songs' Track01.mp3:0 Track02.mp3
.TP 
\fBA subject line from the above post may look like this:\fP
.br 
2 songs \- File 2 of 2: Track02.mp3 (1/8)
.TP 
\fBPost an .SFV file and Track02.mp3, but include Track01.mp3 in the .SFV file:\fP
.br 
newspost \-c songs.SFV \-s '2 songs' Track01.mp3:0 Track01.mp3
.ad
.SH "AUTHORS"
.LP 
.ad
Mostly written by Jim Faulkner <newspost@unixcab.org>, with some help from 
William McBrine <wmcbrine@users.sf.net>. Portions borrowed from cksfv by 
Bryan Call, and parchive by Willem Monsuwe.
.SH "AVAILABILITY"
The most recent version of newspost is available at
http://newspost.unixcab.org/


CHANGES
Newspost 

List of Changes

1.12
 - file permissions on ~/.newspost are now 600
 - fixed bug in creating headers when -t option is used
 - homepage is now included in the user-agent line
 - new -V switch to check the version

1.11
 - the number of digits in the part number field matches the number
of digits in the total part count (thanks Mike Liang)

1.1
 - no longer uses uulib
 - added a makefile for solaris
 - now can reply to messages using -r
 - misc bugfixes

1.01a
 - fixed typo in newspost -h

1.01
 - fixed problems with some news servers

1.0
 - initial release



newspost.c

/* Newspost 1.13

   Copyright (C) 2000 Jim Faulkner <jfaulkne@ccs.neu.edu>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA. */

#include "newspost.h"

struct filell *parseargs(int argc, char **argv, 
               struct headerinfo *header, 
               char *address, int *port, 
               char *username, char *password,
               bool *istext, bool *setasdefault,
	       int *maxlines, struct filell **zerofile,
               char *sfvname);

void printhelp(struct headerinfo *header,
               char *address, int *port,
               char *username, char *password,
               int *maxlines);

void needarg(const char *myarg);

bool writedefaults(struct headerinfo *header,
                   char *address,int *port,
                   char *username,char *password,
                   int *maxlines);

void getdefaults(struct headerinfo *header,
                 char *address,int *port,
                 char *username,char *password,
                 int *maxlines);

extern struct filell *checkfiles(struct filell *firstfile);

extern struct filell *addfile(const char *myfile, struct filell *first);

extern struct filell *freefilell(struct filell *first,struct filell *tofree);

char *reference = NULL;

int main(int argc, char **argv){
  struct postsocket sock;
  int port;
  int maxlines;
  struct headerinfo header;
  char unstripped[1024];
  bool istext = FALSE;
  bool setasdefault = FALSE;
  float totalsize = 0;
  int numfiles = 0;
  int begintime = 0;
  float totaltime = 0;
  struct filell *firstfile = NULL;
  struct filell *loc = firstfile;
  struct filell *zerofile = NULL;
  struct filell *oldfirstfile = NULL;
  char *sfvname = calloc(1024,sizeof(char)); 
  char *address = calloc(1024,sizeof(char));
  char *username = calloc(1024,sizeof(char));
  char *password = calloc(1024,sizeof(char));
  header.from = calloc(1024,sizeof(char));
  header.newsgroups = calloc(1024,sizeof(char));
  header.subject = calloc(1024,sizeof(char));
  header.organization = calloc(1024,sizeof(char));
  header.useragent = calloc(1024,sizeof(char));

  getdefaults(&header,address,&port,username,password,&maxlines);

  firstfile = parseargs( argc, argv, &header, 
                         address, &port, 
                         username, password, 
                         &istext, &setasdefault, 
                         &maxlines, &zerofile,
                         sfvname );

  if(setasdefault == TRUE){ 
    if(writedefaults(&header,
                  address,&port,
                  username,password,&maxlines)==FALSE){
      printf("Error writing ~/.newspost\n");
    }
    else{
      printf("Defaults written to ~/.newspost\n");
    }
  }

  firstfile = checkfiles(firstfile);

  zerofile = checkfiles(zerofile);

  if(firstfile == NULL){
    printf("No files to post!\n");
    exit(0);
  }
  if( (istext == TRUE) 
      && (firstfile->next != NULL) ){
    printf("Only one text file may be posted at a time.\n");
    exit(0);
  }
  if( (istext == TRUE)
      && (zerofile != NULL) ){
    printf("Only one text file may be posted at a time.\n");
    exit(0);
  }
  if( (header.subject == NULL)
      || (strncmp(header.subject,"Subject: ",9)!=0)){
    printf("There is no subject line.\n");
    exit(0);
  }
  if( (header.organization == NULL)
      || (strncmp(header.organization,"Organization: ",14)!=0)){
    printf("There is no organization line.\n");
    exit(0);
  }
  if( (header.from == NULL)
      || (strncmp(header.from,"From: ",6)!=0)){
    printf("There is no from line.\n");
    exit(0);     
  }
  if( (header.newsgroups == NULL)
      || (strncmp(header.newsgroups,"Newsgroups: ",12)!=0)){
    printf("There is no newsgroups line.\n");
    exit(0);
  }
  if( (address == NULL)
      || (strcmp(address,"")==0)
      || (strcmp(address,"\n")==0)
      || (strcmp(address,"\r\n")==0)){
    printf("You must provide a server name or set the NNTPSERVER environment variable.\n");
    exit(0);
  }

  loc = firstfile;
  while(loc != NULL){
    totalsize += loc->fileinfo.st_size;
    numfiles++;
    loc = loc->next;
  }

  printf("\n");
  printf("%s",header.from);
  printf("%s",header.newsgroups);
  printf("%s\n",header.subject);
  printf("%s",header.organization);
  if(reference != NULL) printf("%s",reference);
  printf("\n");
  if(istext != TRUE){
    if(totalsize < (1024*1024))
      printf("Posting %.0f bytes in %d files.\n",totalsize,numfiles);
    else
      printf("Posting %.1f megabytes in %d files.\n",((totalsize/1024)/1024),numfiles);
  }
  else
    printf("Posting %s as body of message.\n",firstfile->filename);
  printf("Posting will begin in %d seconds.\n",SLEEPTIME);
  sleep(SLEEPTIME);
  printf("\n");

  socket_create(&sock,address,port);
  socket_getline(&sock);
  /* debug1("%s",sock.line_buffer); */

  if( ((username != NULL) 
      && (strcmp(username,"")!=0)
      && (strcmp(username,"\n")!=0)
      && (strcmp(username,"\r\n")!=0))
     ||
      ((password != NULL)
      && (strcmp(password,"")!=0)
      && (strcmp(password,"\n")!=0)
      && (strcmp(password,"\r\n")!=0))){
    if(nntpauth(&sock,username,password)==FALSE){
      printf("Incorrect username or password.\n");
      exit(0);
    }
  }

  if(istext == TRUE){
    printf("Posting %s as text.\n",firstfile->filename);
    sprintf(header.subject,"%s\r\n",header.subject);
    postfile(&header,firstfile->filename,&sock,NULL,NULL);
    printf("\n");
    printf("Posted %s\n",firstfile->filename);
  }
  else{
    totalsize = 0;
    oldfirstfile = firstfile;
    while(firstfile != NULL){
      begintime = time(0);
      encode_and_post(&sock,firstfile->filename,&header,maxlines,zerofile);
      zerofile = NULL;
      begintime = (time(0) - begintime);
      totaltime += begintime;
      totalsize += firstfile->fileinfo.st_size;
      if(begintime != 0)
	printf("Posted %s (%d bytes) in %d seconds (%d KB/second)\n",firstfile->filename,firstfile->fileinfo.st_size,begintime, ((firstfile->fileinfo.st_size/begintime)/1024) );
      else
	printf("Posted %s (%d bytes) in 0 seconds (very quickly)\n",firstfile->filename,firstfile->fileinfo.st_size);
      firstfile = firstfile->next;
    }

    if(sfvname != NULL
       && (strcmp(sfvname,"")!=0)
       && (strcmp(sfvname,"\n")!=0)
       && (strcmp(sfvname,"\r\n")!=0)){
      sprintf(unstripped,"/tmp/%s",sfvname);
      newsfv(oldfirstfile,unstripped);
      encode_and_post(&sock,unstripped,&header,maxlines,zerofile);
      unlink(unstripped); 
      printf("Posted %s\n",sfvname);
    }

    printf("\n");
    if(totaltime != 0)
      printf("Posted %d files.\nTotal %.0f bytes in %.0f minutes (%.0f KB/second)\n",numfiles,totalsize,(totaltime/60),((totalsize/totaltime)/1024));
    else
      printf("Posted %d files.\nTotal %.0f bytes in 0 minutes (really quickly)\n",numfiles,totalsize);
  }
  /* This stuff has yet to be free()'d */
  /* but hey, we're exiting anyway */
  /*
  while(oldfirstfile != NULL){
    oldfirstfile = freefilell(oldfirstfile,oldfirstfile);
  }
  if(reference != NULL){
    free(reference);
  }
  free(password);
  free(address);
  free(username);
  free(header.subject);
  free(header.newsgroups);
  free(header.from);
  free(header.organization);
  free(header.useragent);
  */
  return 0;
}

struct filell *parseargs(int argc, char **argv, 
               struct headerinfo *header, 
               char *address, int *port, 
               char *username, char *password,
               bool *istext, bool *setasdefault,
               int *maxlines, struct filell **zerofile,
               char *sfvname){
  struct filell *firstfile = NULL;
  int i = 1;
  if(argc==1) printhelp(header,address,port,
                        username,password,maxlines);
  while(i<argc){
    if(strcmp(argv[i],"-s")==0){
      /* get the subject line */
      if((i+1)==argc) needarg("-s");
      i++;
      sprintf(header->subject,"Subject: %s",argv[i]);
    }
    else if(strcmp(argv[i],"-n")==0){
      /* get the newsgroups line */
      if((i+1)==argc) needarg("-n");
      i++;
      sprintf(header->newsgroups,"Newsgroups: %s\r\n",argv[i]);
    }
    else if( (strcmp(argv[i],"-h")==0)
            || (strcmp(argv[i],"-?")==0))
	   printhelp(header,address,port,
                     username,password,maxlines);
    else if(strcmp(argv[i],"-f")==0){
      /* get the from line */
      if((i+1)==argc) needarg("-f");
      i++;
      sprintf(header->from,"From: %s\r\n",argv[i]);
    }
    else if(strcmp(argv[i],"-o")==0){
      /* get the organization line */
      if((i+1)==argc) needarg("-o");
      i++;
      sprintf(header->organization,"Organization: %s\r\n",argv[i]);
    }
    else if(strcmp(argv[i],"-d")==0){
      /* set as the default */
      *setasdefault = TRUE;
    }
    else if(strcmp(argv[i],"-t")==0){
      /* post a text message */
      *istext = TRUE; 
    }
    else if(strcmp(argv[i],"-u")==0){
      /* get the username */
      if((i+1)==argc) needarg("-u");
      i++;
      strcpy(username,argv[i]);
    }
    else if(strcmp(argv[i],"-p")==0){
      /* get the password */
      if((i+1)==argc) needarg("-p");
      i++;
      strcpy(password,argv[i]);
    }
    else if(strcmp(argv[i],"-r")==0){
      /* get the referenced message id */
      if((i+1)==argc) needarg("-r");
      reference = calloc(1024,sizeof(char));
      i++;
      sprintf(reference,"Reference: %s\r\n",argv[i]);
    }
    else if(strcmp(argv[i],"-i")==0){
      /* get the server name or ip address */
      if((i+1)==argc) needarg("-i");
      i++;
      strcpy(address,argv[i]);
    }
    else if(strcmp(argv[i],"-z")==0){
      /* get the port number */
      if((i+1)==argc) needarg("-z");
      i++;
      *port = atol(argv[i]);
    }
    else if(strcmp(argv[i],"-l")==0){
      /* get the maximum number of lines */
      if((i+1)==argc) needarg("-l");
      i++;
      *maxlines = atol(argv[i]);
    }
    else if(strcmp(argv[i],"-0")==0){
      /* get the zero file */
      if((i+1)==argc) needarg("-0");
      i++;
      *zerofile = addfile(argv[i],*zerofile);
    }
    else if(strcmp(argv[i],"-v")==0){
      /* get the sfv file */
      if((i+1)==argc) needarg("-v");
      i++;
      strcpy(sfvname,argv[i]);
    }
    else if((strcmp(argv[i],"-V")==0)){
      /* display the version information and exit */
      printf("newspost v%s\n",NVERSION);
      exit(0);
    }
    else{
      firstfile = addfile(argv[i], firstfile); 
    }
    i++;
  }
  return firstfile; 
}

void printhelp(struct headerinfo *header,
               char *address, int *port,
               char *username, char *password,
               int *maxlines){
  printf("newspost v%s\n",NVERSION);
  printf("Usage: newspost [OPTIONS [ARGUMENTS]] file1 file2 file3...\n");
  printf("Options:\n");
  printf("  -h, -?  - prints help.\n");
  printf("  -s SUBJ - subject for all of your posts.\n");
  if( (header->newsgroups != NULL)
      && (strcmp(header->newsgroups,"")!=0)
      && (strcmp(header->newsgroups,"\n")!=0)
      && (strcmp(header->newsgroups,"\r\n")!=0)){
    printf("  -n NWSG - newsgroups to post to, seperated by commas. Default %s",header->newsgroups);
  }
  else printf("  -n NWSG - newsgroups to post to, seperated by commas.\n");

  if( (header->from != NULL)
      && (strcmp(header->from,"")!=0)
      && (strcmp(header->from,"\n")!=0)
      && (strcmp(header->from,"\r\n")!=0)){
    printf("  -f FROM - your e-mail address. Default %s",header->from); 
  }
  else printf("  -f FROM - your e-mail address.\n");
  if( (header->organization != NULL)
      && (strcmp(header->organization,"")!=0)
      && (strcmp(header->organization,"\n")!=0)
      && (strcmp(header->organization,"\r\n")!=0)){
    printf("  -o ORGN - your organization. Default %s",header->organization);
  }
  else printf("  -o ORGN - your organization.\n");
  if( (address != NULL)
      && (strcmp(address,"")!=0)
      && (strcmp(address,"\n")!=0)
      && (strcmp(address,"\r\n")!=0)){
    printf("  -i ADDR - hostname or ip address of your news server. Default Server: %s\n",address);
  }
  else printf("  -i ADDR - hostname or ip address of your news server.\n");
  printf("  -z PORT - port number of the NNTP server if nonstandard. Default Port: %d\n",*port);
 
  if( (username != NULL)
      && (strcmp(username,"")!=0)
      && (strcmp(username,"\n")!=0)
      && (strcmp(username,"\r\n")!=0)){
    printf("  -u USER - username, only used if authorization needed. Default Username: %s\n",username);
  }
  else printf("  -u USER - username, only used if authorization needed.\n");
  
  if( (password != NULL)
      && (strcmp(password,"")!=0)
      && (strcmp(password,"\n")!=0)
      && (strcmp(password,"\r\n")!=0)){
    printf("  -p PASS - password, only used if authorization needed. Default Password: ********\n",password); 
  }
  else printf("  -p PASS - password, only used if authorization needed.\n");
  printf("  -l LNES - maximum number of lines per post. Default Maximum Lines: %d\n",*maxlines);
  printf("  -0 FILE - FILE is posted as text as a prefix (0/xx) to the first file posted.\n");
  printf("  -v NAME - create and post as a sfv checksum file.\n");
  printf("  -d      - set current options as your default.\n");
  printf("  -t      - post as text (no uuencoding). only one file may be posted.\n");
  printf("  -r MGID - reply to MESSAGE_ID\n");
  printf("  -V      - display version information and exit\n");
  printf("Examples:\n");
  printf("  $ newspost -f you@yourbox.yourdomain -o \"Your Organization\" -i news.yourisp.com -n alt.binaries.sounds.mp3 -d\n");
  printf("  $ newspost -s \"This is me singing\" -0 musicinfo.txt *.mp3 -v mysongs.sfv\n");
  printf("  $ newspost -s \"Some pics of my kitty\" -n alt.binaries.pets.cats kitty1.jpg kitty2.jpg kitty3.jpg\n");
  exit(0);
}

void needarg(const char *myarg){
  printf("The %s option requires an argument.\n",myarg);
  exit(0);
}

bool writedefaults(struct headerinfo *header,
                   char *address,int *port,
                   char *username,char *password,
                   int *maxlines){
  char defaultsfile[1024];
  char tmp[1024];
  FILE *myfile;
  char *homedir;

  homedir = getenv("HOME");
  if(homedir != NULL) sprintf(defaultsfile,"%s/.newspost",homedir);
  else sprintf(defaultsfile,"~/.newspost");
  myfile = fopen(defaultsfile,"w");
  if (myfile == NULL) return FALSE;
  sprintf(tmp,"%s",(header->from+6));
  if(header->from[1] != 0) fprintf(myfile,"%s",tmp);
  else fprintf(myfile,"\r\n");
  sprintf(tmp,"%s",(header->organization+14));
  if(header->organization[1] != 0) fprintf(myfile,"%s",tmp);
  else fprintf(myfile,"\r\n");
  sprintf(tmp,"%s",(header->newsgroups+12));
  if(header->newsgroups[1] != 0) fprintf(myfile,"%s",tmp);
  else fprintf(myfile,"\r\n");
  fprintf(myfile,"%s\r\n",address);
  fprintf(myfile,"%d\r\n",*port);
  fprintf(myfile,"%s\r\n",username);
  fprintf(myfile,"%s\r\n",password);
  fprintf(myfile,"%d\r\n",*maxlines);
  fclose(myfile);
  chmod(defaultsfile,00600);
  return TRUE;
}

void getdefaults(struct headerinfo *header,
                 char *address,int *port,
                 char *username,char *password,
                 int *maxlines){
  char *homedir;
  char defaultsfile[1024];
  FILE *myfile;
  char thisline[1024];
  int i = 0;
  int j = 0;
  char *loc;
  char *envvar = NULL;
  char *tmp1 = NULL;
  char *tmp2 = NULL;

  sprintf(header->useragent,"User-Agent: newspost v%s http://www.ccs.neu.edu/home/jfaulkne/newspost/\r\n",NVERSION);
  sprintf(header->organization,"Organization: %s\r\n",ORGANIZATION);
  sprintf(header->subject,"");
  sprintf(header->from,"");
  sprintf(header->newsgroups,"");
  envvar = getenv("NNTPSERVER");
  if(envvar == NULL) sprintf(address,"");
  else strcpy(address,envvar);
  tmp1 = getenv("USER");
  tmp2 = getenv("USERNAME");
  envvar = getenv("HOSTNAME");
  if( (tmp1 != NULL)
      && (envvar != NULL)){
    if(tmp2 != NULL)
      sprintf(header->from,"From: %s <%s@%s>\r\n",tmp2,tmp1,envvar);
    else
      sprintf(header->from,"From: %s@%s\r\n",tmp1,envvar);
  }
  *port = 119;
  *maxlines = 7500;

  homedir = getenv("HOME");
  if(homedir != NULL) sprintf(defaultsfile,"%s/.newspost",homedir);
  else sprintf(defaultsfile,"~/.newspost");
  myfile = fopen(defaultsfile,"r");
 
  if(myfile != NULL){
    while(i<8){
      fgets(thisline,1024,myfile);
      if( (strcmp(thisline,"")!=0)
         && (strcmp(thisline,"\n")!=0)
         && (strcmp(thisline,"\r\n")!=0)){

        if(i==0) sprintf(header->from,"From: %s",thisline);
        else if(i==1) sprintf(header->organization,"Organization: %s",thisline);
        else if(i==2) sprintf(header->newsgroups,"Newsgroups: %s",thisline);
        else if(i==3){
          loc = thisline;
          while( (*loc != '\r') 
                && (*loc != '\n')){
            loc++;
            j++;
          }
          strncpy(address,thisline,j);
        }
        else if(i==4) *port = atol(thisline);
        else if(i==5){
          j = 0;
          loc = thisline;
          while( (*loc != '\r')
                && (*loc != '\n')){
            loc++;
            j++;
          }
          strncpy(username,thisline,j);
        } 
        else if(i==6){
          j = 0;
          loc = thisline;
          while( (*loc != '\r')
                && (*loc != '\n')){
            loc++;
            j++;
          }
          strncpy(password,thisline,j);
        }
        else if(i==7) *maxlines = atol(thisline);
      }
      i++;
    }
    fclose(myfile);
  }
}

